Code/Tools/AssetProcessor/native/resourcecompiler/rcjob.cpp

```diff
@@ -828,9 +831,28 @@ namespace AssetProcessor
                     relativeFilePath = product.m_outputPathOverride;
                 }
 
-                if (!VerifyOutputProduct(
-                    QDir(intermediateDirectory.c_str()), outputFilename, absolutePathOfSource,
-                    fileSizeRequired, outputsToCopy))
+                if (VerifyOutputProduct(
+                        QDir(intermediateDirectory.c_str()), outputFilename, absolutePathOfSource, fileSizeRequired, outputsToCopy))
+                {
+                    // A null uuid indicates the top level source is not using metadata files.
+                    // The assumption for the UUID generated below is that the source UUID will not change.  A type which has no metadata
+                    // file currently may be updated later to have a metadata file, which would break that assumption.  In that case, stick
+                    // with the default path-based UUID.
+                    if (!params.m_topLevelSourceUuid.IsNull())
+                    {
+                        // Generate a UUID for the intermediate as:
+                        // SourceUuid:BuilderUuid:SubId
+                        auto uuid = AZ::Uuid::CreateName(AZStd::string::format(
+                            "%s:%s:%d",
+                            params.m_topLevelSourceUuid.ToFixedString().c_str(),
```

Comment:
Should we use the top level source UUID or the UUID of the direct source here? For example, consider a processing chain like this: source asset A -> intermediate asset B -> intermediate asset C -> product asset B. The UUID of intermediate asset C should be SourceUuidOfA:FirstBuilderUuid:SecondBuilderUuid:SubIdOfC instead of SourceUuidOfA:SecondBuilderUuid:SubIdOfC.

Reply:
If we include the UUIDs of all the builders in the chain it would make it difficult for users (asset builders) to generate this UUID.  Having the last builder + subID should be easy enough as those should be knowable.  As for the UUID of the top level source vs the direct source... I originally thought the top level would be easier to figure out for a builder trying to create a reference, but now I'm questioning that.  We might need to spend some time to figure out which is best here

Reply:
Ok so I spent some time thinking about this and I think using the top level source is not what we want here.   Here's the scenario I'm thinking about:

- FBX file generates material intermediate
- TXT file generates FBX file intermediate which generates material intermediate

Python file is attached to FBX file and wants to set up a reference to the material file.  If the ID is generated from the top level source, it will be difficult for the python file to find out the TXT file asset ID.  If it's generated from the FBX file (the direct source) it should be easy to look up and calculate.